pkg/secrets/keyvault/client.go

```diff
@@ -307,55 +309,112 @@ func (k *KeyvaultSecretClient) Delete(ctx context.Context, key types.NamespacedN
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if options.Flatten {
+		for _, suffix := range options.FlattenSuffixes {
+			secretName, err := k.makeSecretName(secrets.SecretKey{Name: key.Name + "-" + suffix, Namespace: key.Namespace, Kind: key.Kind})
+			if err != nil {
+				return err
+			}
+			err = k.deleteKeyVaultSecret(ctx, secretName)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
 	} else {
-		secretName = key.Name
+		secretName, err := k.makeSecretName(key)
+		if err != nil {
+			return err
+		}
+		return k.deleteKeyVaultSecret(ctx, secretName)
 	}
+}
 
-	secretVersion := ""
+// Get gets a key from KeyVault
+func (k *SecretClient) Get(ctx context.Context, key secrets.SecretKey, opts ...secrets.SecretOption) (map[string][]byte, error) {
+	options := &secrets.Options{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	vaultBaseURL := GetVaultsURL(k.KeyVaultName)
 	data := map[string][]byte{}
 
+	secretName, err := k.makeSecretName(key)
+	if err != nil {
+		return data, err
+	}
+
+	secretVersion := ""
 	result, err := k.KeyVaultClient.GetSecret(ctx, vaultBaseURL, secretName, secretVersion)
 
 	if err != nil {
-		return data, fmt.Errorf("secret does not exist" + err.Error())
+		return data, errors.Wrapf(err, "secret %q could not be found in KeyVault %q", secretName, vaultBaseURL)
 	}
 
 	stringSecret := *result.Value
 
-	// Convert the data from json string to map
-	jsonErr := json.Unmarshal([]byte(stringSecret), &data)
-
-	// If Unmarshal fails on the input data, the secret likely not a json string so we return the string value directly rather than unmarshaling
-	if jsonErr != nil {
-		data = map[string][]byte{
-			secretName: []byte(stringSecret),
+	// If flatten is enabled, we're getting a single secret entry and so it won't be JSON encoded
+	if options.Flatten {
+		// This is quite hacky -- it's only used in test though...
+		data["secret"] = []byte(stringSecret)
+	} else {
+		// Convert the data from json string to map
+		err = json.Unmarshal([]byte(stringSecret), &data)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to deserialize secret %q in KeyVault %q", secretName, vaultBaseURL)
 		}
 	}
 
 	return data, err
 }
 
-// Create creates a key in KeyVault if it does not exist already
-func (k *KeyvaultSecretClient) CreateEncryptionKey(ctx context.Context, name string) error {
-	vaultBaseURL := getVaultsURL(ctx, k.KeyVaultName)
-	var ksize int32 = 4096
-	kops := keyvault.PossibleJSONWebKeyOperationValues()
-	katts := keyvault.KeyAttributes{
-		Enabled: to.BoolPtr(true),
-	}
-	params := keyvault.KeyCreateParameters{
-		Kty:           keyvault.RSA,
-		KeySize:       &ksize,
-		KeyOps:        &kops,
-		KeyAttributes: &katts,
-	}
-	k.KeyVaultClient.CreateKey(ctx, vaultBaseURL, name, params)
-	return nil
+func (k *SecretClient) makeLegacySecretName(key secrets.SecretKey) (string, error) {
+	// The legacy behavior had differences between different kinds of resources - we need to preserve those differences
```

Comment:
Vestigial from when you were going to do all of the name generation here?

Reply:
Oh, yep. Removed.